tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -101,9 +101,12 @@ ctorForProductType constructorName fields = recC (toName Named constructorName)
   toVarBangType name (MkField required fieldTypes mult) =
     let fieldName = mkName . addTickIfNecessary . removeUnderscore $ name
         strictness = TH.bang noSourceUnpackedness noSourceStrictness
-        contents = if required == Optional then conT ''Maybe `appT` choices else choices
         ftypes = fieldTypesToNestedEither fieldTypes
-        choices = if mult == Multiple then appT (conT ''[]) ftypes else ftypes
+        contents = case (required, mult) of
+          (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
+          (Required, Single) -> ftypes
+          (Optional, Multiple) -> appT (conT ''[]) ftypes
+          (Optional, Single) -> appT (conT ''Maybe) ftypes
```

Comment:
This truth table represents the four combinations we could have and build different data types accordingly.